### DISCUSSION TITLE
CPU load causes tautomer canonicalization to exit early

### DISCUSSION THREAD

**Discussion by pechersky** [rdkit/rdkit#7966] [Q&A]
We have a web server running that, among other things, does tautomer canonicalization. Somehow, when under load, the canonicalization exits early with messages like
```
Tautomer enumeration stopped at 20 tautomers: max tautomers reached
Tautomer enumeration stopped at 8 tautomers: max transforms reached
```

When trying those molecules again locally, or later when the server is less loaded, they're fine.
Our code path looks like below. I haven't made an issue about this because I can't recreate this locally, even with all my attempts with `cputool` or cpu limiting in a docker-compose file. I also read the canonicalizer source and can't find where high load would cause an early exit. Any ideas? I'm stumped.

```
from rdkit import Chem
from rdkit.Chem.MolStandardize import rdMolStandardize
from rdkit import RDLogger

lg = RDLogger.logger()
lg.setLevel(RDLogger.CRITICAL)

rdkit_mol = Chem.MolFromMolFile("foo.sdf")
enumerator.SetMaxTautomers(20)
enumerator.SetMaxTransforms(20)
enumerator.SetRemoveBondStereo(False)
enumerator.SetRemoveSp3Stereo(False)
canon_mol = enumerator.Canonicalize(rdkit_mol)
```

Here's an example showing the load
![image](https://github.com/user-attachments/assets/b3930796-5e31-42cb-a1f7-2cfd51f880a9)

**Comment by bp-kelley** [comment]
Tautomerization halting is a timeout, not a # of cycles so heavy CPU utilization will impact the number of tautomer steps.  

For tautomer canonicalization, I believe this is a bug, there should be a max # tautomers generated and not a timeout.  Otherwise we can't guarantee canonicalization.

**Reply by pechersky** [reply]
Right, that's why we raise the error when we see the log lines instead of returning the partial result. I can't pin down the codepath that exits on timeout, though. I see there is the TautomerEnumeratorCallback and how  python tests set their own, but not where the default callback is defined.

**Reply by bp-kelley** [reply]
It’s buried in the tautomerenumerator, it has a return code that specifies if there was a timeout

**Reply by pechersky** [reply]
Thanks for the explanation. We've increased the max transform+tautomer cap, overrode the callback, and also uncovered a thread race condition with `WrapLogs`!

**Comment by pechersky** [comment]
> For tautomer canonicalization, I believe this is a bug, there should be a max # tautomers generated and not a timeout.

Bumping the values to 10K and removing the timeout still ends in a balk on something like cyclosporine, unfortunately.
